### ISSUE TITLE
Introduction/LastPush Task 4/8 doesn't pass a test due to line separator

### ISSUE THREAD

**Issue by SergeiPavlitz** [jetbrains-academy/kotlin-onboarding-introduction#230]
Introduction/LastPush Task 4/8: Patterns Generator - repeatHorizontally Function, doesn't pass due to line separator.

After implementation of repeatHorizontally function, some test doesn't pass.
The result of my function and expected result are the same.
It seems the problem in expected line separator.
Could you explain please, how to fix or bypass the issue?

My function:
```
fun repeatHorizontally (pattern: String, n: Int, patternWidth: Int): String {
    val result = StringBuilder()
    pattern.lines().forEach { s: String ->
        val filled = fillPatternRow(s, patternWidth)
        val multiString = filled.repeat(n)
        result.append(multiString).append(newLineSymbol)
    }
    result.removeSuffix(newLineSymbol)
    return result.toString()
}
```

Feedback and the comparison  result you can find in the attachments.

**OS**: Windows 10 home edition
**IDE**: IntelliJ IDEA 2023.2.5 (Ultimate Edition); Build #IU-232.10227.8, built on November 9, 2023; Runtime version: 17.0.9+7-b1000.46 amd64; VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.

[feedback.txt](https://github.com/jetbrains-academy/kotlin-onboarding-introduction/files/15309744/feedback.txt)
![testResults 2024-05-14 160650](https://github.com/jetbrains-academy/kotlin-onboarding-introduction/assets/114855170/a1ec40eb-bb53-40a3-9417-3d15a56e06f2)

**Comment by Vigo61** [comment]
I have a similar problem. I can pass Task 4/8 with this function

```
fun repeatHorizontally(pattern: String, n: Int, patternWidth: Int): String {
    val rows = pattern.lines()
    val sb = StringBuilder()
    for ((index,row) in rows.withIndex()) {
        sb.append(fillPatternRow(row, patternWidth).repeat(n))
        if (index != rows.lastIndex )
            sb.append("\n")    }
    return sb.toString()
}
```
Note the use of "\n" instead of newLineSymbol = System.lineSeparator() (\r\n in windows)

I've cloned this repository but I still get similar errors (in test.zip attachment) related to tast 8/8
[test.zip](https://github.com/jetbrains-academy/kotlin-onboarding-introduction/files/15502925/test.zip)

**Comment by nbirillo** [comment]
@SergeiPavlitz @Vigo61 Please, check now, it should work

**Comment by SergeiPavlitz** [comment]
Hi.
I've checked again. The situation is the same.(

But I could bypass the problem, in a way which was described by @Vigo61. I've changed the newLineSymbol on "\n".

I reopened the project via the [website](https://plugins.jetbrains.com/plugin/21067-kotlin-onboarding-introduction) and clicked the button "Open in Intellij Idea".
Maybe this information can help.

My code:

```
fun fillPatternRow(patternRow: String, patternWidth: Int): String {
    check(patternRow.length <= patternWidth) { "The row length is bigger than patternWidth" }
    val extendedRow = StringBuilder(patternRow)
    extendedRow.append(separator.toString().repeat(patternWidth - patternRow.length))
    return extendedRow.toString()
}

fun repeatHorizontally(pattern: String, n: Int, patternWidth: Int): String {
    val result = StringBuilder()
    val list = pattern.lines().toMutableList()
    val first = list.removeFirst()
    result.append(fillPatternRow(first, patternWidth).repeat(n))

    list.forEach { s: String ->
        result.append(newLineSymbol)
//        result.append("\n")
        val filled = fillPatternRow(s, patternWidth)
        val multiString = filled.repeat(n)
        result.append(multiString)
    }
    return result.toString()
}
```


[feedback.txt](https://github.com/user-attachments/files/15585844/feedback.txt)
![screen 2024-06-05 125445](https://github.com/jetbrains-academy/kotlin-onboarding-introduction/assets/114855170/73930fe7-ccc8-4b76-8bca-44d2827c5ae7)

**Comment by Vigo61** [comment]
Hi Sergei,
my version of repeatHorizontally works fine only in task 4 - It's better to use the following function that use newLineSymbol, 
taken from the present git
```
fun repeatHorizontally(pattern: String, n: Int, patternWidth: Int): String {
    val pictureRows = pattern.lines()
    val sb = StringBuilder()
    for (row in pictureRows) {
        val currentRow = fillPatternRow(row, patternWidth)
        sb.append(currentRow.repeat(n))
        sb.append(newLineSymbol)
    }
    return sb.toString()
}
```
but unluckily in the next tasks I get errors from
```
fun dropTopFromLine(line: String, width: Int, patternHeight: Int, patternWidth: Int): String {
    val nToDrop = if (patternHeight > 1) {
        patternWidth * width + newLineSymbol.length
    } else {
        0
    }
    val newPattern = line.removeSuffix(newLineSymbol).drop(nToDrop)
    return "$newPattern$newLineSymbol"
} // this function is not mine still from this git 

```
In this case the error is understandable and I'm trying to correct by myself  before opening a new issue, but it need to me a lot of time :-) 

```
The method dropTopFromLine with arguments line=
+---+---+
| o   o |
|   ^   |
|  ---  |
+---+---+
, width=1, patternHeight=5, patternWidth=9 should return
| o   o |
|   ^   |
|  ---  |
+---+---+
But it returns
o   o |
|   ^   |
|  ---  |
+---+---+
```

**Comment by Vigo61** [comment]
@Byte-m-e 
> I have used `return sb.toString().trimIndent()` instead of `return sb.toString()` and this solved the problem for me.

Which task are you talking about, always of 4) or the later?

Many different solution seems to work fine trying to run manually the main function, but i think there is still some error in the test process.

In can pass tasks from 5) to 7) using this function improved with your suggestion (added .trimIndent())
(derived trom repeatHorizzontally, only the first parameter is needed, the others added to pass the test - so this couldn't be the right solution) 
```
fun dropTopFromLine(line: String, width: Int, patternHeight: Int, patternWidth: Int): String {
    if (patternHeight < 2) return line
    val lineRows = line.lines()
    val sb = StringBuilder()
    for (r in lineRows.indices) {
        if (r == 0) continue
        sb.append(lineRows[r])
        sb.append(newLineSymbol)
    }
    return sb.toString().trimIndent()
}
```
but in task 8) the best function is the original from @nbirillo (from this repository)

```
fun dropTopFromLine(line: String, width: Int, patternHeight: Int, patternWidth: Int): String {
    val nToDrop = if (patternHeight > 1) {
        patternWidth * width + newLineSymbol.length
    } else {
        0
    }
    val newPattern = line.removeSuffix(newLineSymbol).drop(nToDrop)
    return "$newPattern$newLineSymbol"
}
```
but not enough good to pass the whole test process (passed 145 of 147 tests)

_The following errors happen only during test process - because dropTopFromLine is called only from canvasGenerator and the pattern is passed vie repeatHorizontally_

```
fun canvasGenerator(pattern: String, width: Int, height: Int): String {
    val patternWidth = getPatternWidth(pattern)
    val patternHeight = getPatternHeight(pattern)
    val repeatedPattern = repeatHorizontally(pattern, width, patternWidth)
    val dropped = dropTopFromLine(repeatedPattern, width, patternHeight, patternWidth)
    return "$repeatedPattern${baseGenerator(dropped, dropped, height - 1)}"
}
```

```
The method dropTopFromLine with arguments line=
+---+---+
| o   o |
|   ^   |
|  ---  |
+---+---+
, width=1, patternHeight=5, patternWidth=9 should return
| o   o |
|   ^   |
|  ---  |
+---+---+
But it returns
 o   o |
|   ^   |
|  ---  |
+---+---+

<Click to see difference>

org.opentest4j.AssertionFailedError: The method dropTopFromLine with arguments line=
+---+---+
| o   o |
|   ^   |
|  ---  |
+---+---+
, width=1, patternHeight=5, patternWidth=9 should return
| o   o |
|   ^   |
|  ---  |
+---+---+
But it returns
 o   o |
|   ^   |
|  ---  |
+---+---+
 ==> expected: <| o   o |
|   ^   |
|  ---  |
+---+---+> but was: < o   o |
|   ^   |
|  ---  |
+---+---+>
```

**maybe it's time to open a new issue ....**

**Comment by Vigo61** [comment]
@Byte-m-e Hi
thanks for your suggestions.

I tried the code from this git in a Ubuntu virtual machine. I'm really a dummy user in Linux OS but everything was fine, check passed without errors. I tried the same code in another windows machine and I got the same errors that in my main Windows PC.

**Comment by Vigo61** [comment]
@Byte-m-e Hi
thanks for your suggestions.

I tried the code from this git in a Ubuntu virtual machine. I'm really a dummy user in Linux OS but everything was fine, check passed without errors. I tried the same code in another windows machine and I got the same errors that in my main Windows PC.
 I tried your code snippets and your suggestions but i can't pass the check of Task 8 under Windows 10

Under that OS I get the minimum number of errors (2 out of 147) using without modification the code from this git repository. 

I believe that whoever wrote the test code should modify it so that it does not give different results depending on the operating system.

When I have time and inclination maybe I will do some more tests before opening a new ticket

But at the moment I think I have wasted enough time on this problem although perhaps it could have been a useful exercise.

Thanks again
